core/src/lib.rs

```diff
@@ -23,145 +23,41 @@
-//! `with_upgrade` does **not** implement the `Transport` trait and thus cannot be used as a
-//! transport.
-//!
-//! However the `UpgradedNode` struct returned by `with_upgrade` still provides methods named
-//! `dial`, `listen_on`, and `next_incoming`, which will yield you a `Future` or a `Stream`,
-//! which you can use to obtain the `Output`. This `Output` can then be used in a protocol-specific
-//! way to use the protocol.
-//!
-//! ```no_run
-//! extern crate futures;
-//! extern crate libp2p_ping;
-//! extern crate libp2p_core;
-//! extern crate libp2p_tcp_transport;
-//! extern crate tokio;
-//!
-//! use futures::{Future, Stream};
-//! use libp2p_ping::protocol::Ping;
-//! use libp2p_core::{Transport, upgrade::apply_outbound};
-//! use tokio::runtime::current_thread::Runtime;
-//!
-//! # fn main() {
-//! let ping_dialer = libp2p_tcp_transport::TcpConfig::new()
-//!     // We have a `TcpConfig` struct that implements `Dialer`, and apply a `Ping` upgrade on it.
-//!     .and_then(|socket, _| {
-//!         apply_outbound(socket, Ping::default()).map_err(|e| e.into_io_error())
-//!     })
-//!     // TODO: right now the only available protocol is ping, but we want to replace it with
-//!     //       something that is more simple to use
-//!     .dial("/ip4/127.0.0.1/tcp/12345".parse::<libp2p_core::Multiaddr>().unwrap()).unwrap_or_else(|_| panic!())
-//!     .and_then(|mut pinger| {
-//!         pinger.ping(());
-//!         let f = pinger.into_future().map(|_| ()).map_err(|(e, _)| e);
-//!         Box::new(f) as Box<Future<Item = _, Error = _>>
-//!     });
-//!
-//! // Runs until the ping arrives.
-//! let mut rt = Runtime::new().unwrap();
-//! let _ = rt.block_on(ping_dialer).unwrap();
-//! # }
-//! ```
-//!
-//! ## Grouping protocols
-//!
-//! You can use the `.or_upgrade()` method to group multiple upgrades together. The return value
-//! also implements the `ConnectionUpgrade` trait and will choose one of the protocols amongst the
-//! ones supported.
+//! # General overview
+//!
+//! From a user point of view, the main concepts of libp2p are:
+//!
+//! - A `PeerId` is a unique global identifier for a node on the network. Each node should have a
+//!   different `PeerId`. Normally, a `PeerId` is the hash of the public key used to negotiate
+//!   encryption on the communication channel, thereby guaranteeing that they cannot be spoofed.
+//! - The `Transport` trait define how to reach a remote or listen for incoming remote connections.
+//!   See the `transport` module.
+//! - The `Swarm` struct contains all the active or pending connections to remotes, including all
+//!   the opened protocols.
+//! - The `NetworkBehaviour` trait is implemented on types that can customize the behaviour of the
+//!   `Swarm`. It is the `NetworkBehaviour` that defines what happens on the network.
+//! - The `Topology` trait is implemented on types that can hold the layout of the network.
```

Comment:
Maybe:
```suggestion
//! - The `Topology` trait is implemented for types that need to know about the layout of the network to operate.
```

(I guess my underlying question is: those types that need to know about the surrounding topology actually **must** use the `Topology` trait right? For those cases it's not really optional, which is what `can` sort of implies.)

Reply:
The types that implement `Topology` are just sort of glorified hash maps. You can say that they *hold* the layout of the network. When something needs to know the layout, it gets passed a reference to an implementation of `Topology`.

Reply:
Ok, so how about this: "- The `Topology` trait is implemented for types that hold the layout of a network. When other components need the network layout to operate, they are passed an instance of `Topology`."?